compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp

```diff
@@ -1995,7 +1995,7 @@ extern "C" void LLVMRustContextConfigureDiagnosticHandler(
     std::move(RemarkFile),
     std::move(RemarkStreamer),
     std::move(LlvmRemarkStreamer)
-  ));
+  ), true);
```

Comment:
Yeah I stopped here at seeing that our diagnostics handler wasn’t called anymore, to check whether changes to the PGO diagnostics would be impactful. 

This is great point, thank you, I will look into that.

Reply:
Here's some of what I found. 

On debug builds for helloworld,diesel,exa,cargo,syn together (the latter 2 generate *a lot* of optimization diagnostics, we're talking 6-7GBs in total on cargo for example, that `RespectFilters = true` eliminates), there were 330k diagnostics total. 

They are coming from 2 passes, "sdagisel" the more common (260k, 80%), and "asm-printer" next (72k, 20%).

Here's the top 60% to give you an idea:

```
336578 counts
(  1)   166274 (49.4%, 49.4%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel missed terminator"
(  2)    23695 ( 7.0%, 56.4%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel didn't lower all arguments: ptr"
(  3)     7822 ( 2.3%, 58.8%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel missed"
(  4)     3962 ( 1.2%, 59.9%): optimization diagnostic received, kind: OptimizationAnalysis, pass: "asm-printer", message: "8 instructions in function"
...
```

The rest are all around the <=1% range, with unique fastisel messages mentioning the function itself with missed optimizations.

With filters enabled, LLVM will not do as many of such allocations, and smaller ones. IIUC they start initially small and sometimes when the pass is enabled some additional details are appended. 

In any case:

The asm printer ones are the easiest, they're just optimization analysis diagnostics with the instruction count in a function, coming from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp#L1752-L1755).

I'll point towards a couple of the missed optimizations diagnostics in selection dag isel, but they're going through [`reportFastISelFailure`](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L663-L678) (so the 4 callers will be enough) just before going through `LLVMContext::diagnose` you linked above:
- the most common "FastISel missed terminator"/"FastISel missed" from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L1590-L1608), and this is a case where some details are added on-demand.
- "FastISel didn't lower all arguments: ..." from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L1407-L1413).

(also fun: these passes seem to go through `LegacyPassManager.cpp` which I wasn't expecting to still see :)

It feels like maybe we should land `RespectFilters = true` ourselves, and either open LLVM issues or PRs in the future, what do you think ?

These won't impact max-rss under a good allocator since they're so short-lived, and the improvements on instruction counts locally on cachegrind are understandably small (and on the perfbot it's mostly within noise on syn/cargo).